amplitude.py

```diff
@@ -1,20 +1,79 @@
+import boto3
+import hashlib
+import hmac
 import json
+import os
+import requests
 import sys
 
-def transform(line):
-    try:
-        parsed_line = json.loads(line)
+from threading import Timer
 
-        assert(type(parsed_line) is dict)
+AMPLITUDE_API_KEY = os.environ["FXA_AMPLITUDE_API_KEY"]
+HMAC_KEY = os.environ["FXA_AMPLITUDE_HMAC_KEY"]
 
-        event = parsed_line["Fields"]
+# For crude pre-emptive rate-limit obedience.
+MAX_EVENTS_PER_BATCH = 10
+MAX_BATCHES_PER_SECOND = 100
+BATCH_INTERVAL = 1.0 / MAX_BATCHES_PER_SECOND
 
+def handle (message):
+    # http://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
+    records = json.loads(message)["Records"]
+
+    assert(len(records) == 1)
+
+    record = records[0]
+    if record["eventSource"] != "aws:s3":
+        return
+
+    s3 = boto3.resource("s3", region_name=record["awsRegion"])
+    s3_object = s3.Object(record["s3"]["bucket"]["name"], record["s3"]["object"]["key"])
+
+    send(s3_object.get()["Body"].read().decode("utf-8").splitlines())
+
+def send (events):
+    batch = []
+
+    while len(batch) < MAX_EVENTS_PER_BATCH and len(events) > 0:
+        event = json.loads(events.pop(0))
```

Comment:
How big is `events` likely to be?  I ask because `events.pop(0)` is often an anti-pattern in python, producing O(N^2) behaviour because it pops the first item from the front of the array, then shifts all the other items down one space.  If you need them in order and the list is likely to be non-trivial in length, you could `reverse(events)` and then just `events.pop()` from the end.

Reply:
They don't need to be dispatched in order, that was just my OCD. I'll just dispatch them as they as they arrive instead. IIRC we expect more than 15 thousand events per batch, based on current levels.